File: fastapi/app/schemas/entities.py
from datetime import datetime
from typing import Literal

from pydantic import AliasChoices, BaseModel, ConfigDict, Field

from app.core.limits import DEV_API_LIMIT_PER_MONTH


class UserPreferences(BaseModel):
    showKeyboardEffect: bool = False
    typingSoundEnabled: bool = False
    hashnodeStoreInCloud: bool = False
    hashnodeIntegrated: bool = False
    devtoStoreInCloud: bool = False
    devtoIntegrated: bool = False


class UserDoc(BaseModel):
    userId: str
    displayName: str | None = None
    description: str = ""
    email: str | None = None
    avatarUrl: str | None = None
    username: str
    plan: Literal["free", "pro"] = "free"
    preferences: UserPreferences = Field(default_factory=UserPreferences)
    createdAt: datetime
    updatedAt: datetime


class HashnodeDistribution(BaseModel):
    accessToken: str | None = None
    publicationId: str | None = None


class DevtoDistribution(BaseModel):
    accessToken: str | None = None


class ProjectDoc(BaseModel):
    projectId: str
    ownerId: str
    name: str
    slug: str
    description: str
    contentGuidelines: str = ""
    logoUrl: str | None = None
    isPublic: bool = False
    pagesNumber: int = 0
    createdAt: datetime
    updatedAt: datetime


class CollectionDoc(BaseModel):
    collectionId: str
    projectId: str
    ownerId: str
    name: str
    description: str = ""
    slug: str
    isPublic: bool = False
    pagesNumber: int = 0
    createdAt: datetime
    updatedAt: datetime


class PaperMetadata(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    title: str
    metaDescription: str
    canonical: str
    robots: str
    ogTitle: str
    ogDescription: str
    ogImage: str
    ogImageWidth: int
    ogImageHeight: int
    ogImageAlt: str
    ogLocale: str
    ogPublishedTime: str
    ogModifiedTime: str
    ogAuthorUrl: str
    ogTags: list[str]
    twitterTitle: str
    twitterDescription: str
    twitterImage: str
    twitterImageAlt: str
    twitterCreator: str | None = None
    headline: str
    abstract: str
    keywords: str
    articleSection: str
    wordCount: int
    readingTimeMinutes: int = 1
    inLanguage: str
    datePublished: str
    dateModified: str
    authorName: str
    authorHandle: str = ""
    authorUrl: str
    authorId: str
    coverImageUrl: str
    publisherName: str = "Whitepapper"
    publisherUrl: str = ""
    isAccessibleForFree: bool
    license: str
    # New AI/SEO fields
    keyTakeaways: list[str] | None = None
    faq: list[dict] | None = None
    authorBio: str | None = Field(default=None, validation_alias=AliasChoices("authorBio", "author_bio"))
    jsonLd: dict | list[dict] | None = Field(default=None, validation_alias=AliasChoices("jsonLd", "jsonld"))


class PaperDoc(BaseModel):
    paperId: str
    collectionId: str | None = None
    projectId: str | None = None
    ownerId: str
    thumbnailUrl: str | None = None
    title: str
    slug: str
    body: str = ""
    status: Literal["draft", "published", "archived"] = "draft"
    metadata: PaperMetadata | None = None
    createdAt: datetime
    updatedAt: datetime


class ApiKeySummary(BaseModel):
    keyId: str
    ownerId: str
    projectId: str
    usage: int = 0
    limitPerMonth: int = DEV_API_LIMIT_PER_MONTH
    isActive: bool = True
    createdAt: datetime


class ApiKeyCreateResponse(ApiKeySummary):
    rawKey: str

File: astro/src/lib/entities.ts
export type UserDoc = {
  userId: string;
  displayName?: string | null;
  description: string;
  email?: string | null;
  avatarUrl?: string | null;
  username: string;
  plan: "free" | "pro";
  preferences?: {
    showKeyboardEffect?: boolean;
    typingSoundEnabled?: boolean;
    hashnodeStoreInCloud?: boolean;
    hashnodeIntegrated?: boolean;
    devtoStoreInCloud?: boolean;
    devtoIntegrated?: boolean;
  };
  createdAt: string;
  updatedAt: string;
};

export type HashnodeDistribution = {
  accessToken?: string | null;
  publicationId?: string | null;
};

export type DevtoDistribution = {
  accessToken?: string | null;
};

export type ProjectDoc = {
  projectId: string;
  ownerId: string;
  name: string;
  slug: string;
  description: string;
  contentGuidelines: string;
  logoUrl?: string | null;
  isPublic: boolean;
  pagesNumber: number;
  createdAt: string;
  updatedAt: string;
};

export type CollectionDoc = {
  collectionId: string;
  projectId: string;
  ownerId: string;
  name: string;
  slug: string;
  description: string;
  isPublic: boolean;
  createdAt: string;
  updatedAt: string;
  pagesNumber: number;
};

export type PaperDoc = {
  paperId: string;
  collectionId?: string | null;
  projectId?: string | null;
  ownerId: string;
  thumbnailUrl?: string | null;
  title: string;
  slug: string;
  body: string;
  status: "draft" | "published" | "archived";
  metadata?: PaperMetadata | null;
  createdAt: string;
  updatedAt: string;
};

export type PaperMetadata = {
  title: string;
  metaDescription: string;
  canonical: string;
  robots: string;
  ogTitle: string;
  ogDescription: string;
  ogImage: string;
  ogImageWidth: number;
  ogImageHeight: number;
  ogImageAlt: string;
  ogLocale: string;
  ogPublishedTime: string;
  ogModifiedTime: string;
  ogAuthorUrl: string;
  ogTags: string[];
  twitterTitle: string;
  twitterDescription: string;
  twitterImage: string;
  twitterImageAlt: string;
  twitterCreator?: string | null;
  headline: string;
  abstract: string;
  keywords: string;
  articleSection: string;
  wordCount: number;
  readingTimeMinutes: number;
  inLanguage: string;
  datePublished: string;
  dateModified: string;
  authorName: string;
  authorHandle: string;
  authorUrl: string;
  authorId: string;
  coverImageUrl: string;
  publisherName: string;
  publisherUrl: string;
  isAccessibleForFree: boolean;
  license: string;
  // New AI/SEO fields
  keyTakeaways?: string[] | null;
  faq?: Array<{ question: string; answer: string }> | null;
  author_bio?: string | null;
  jsonld?: Record<string, any> | Array<Record<string, any>> | null;
};
